File: hospital-ai/backend/app/schemas/doctor.py
"""Doctor domain schemas."""

from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Any, List, Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.schemas.patient import PatientGender


class AvailabilityStatus(str, Enum):
    AVAILABLE = "Available"
    BUSY = "Busy"
    ON_LEAVE = "On Leave"


class DoctorAIExtensions(BaseModel):
    """Extension points for future AI agents — currently unused."""

    ai_summary: Optional[str] = None
    performance_metrics: Optional[Any] = None
    predicted_workload: Optional[Any] = None
    recommended_schedule: Optional[Any] = None
    schedule_score: Optional[Any] = None


class DoctorBase(BaseModel):
    first_name: str = Field(min_length=1, max_length=100)
    last_name: str = Field(min_length=1, max_length=100)
    email: EmailStr
    phone: str = Field(min_length=7, max_length=20)
    gender: PatientGender
    date_of_birth: Optional[date] = None
    department_id: Optional[UUID] = None
    specialization: str = Field(min_length=1, max_length=150)
    qualification: Optional[str] = Field(default=None, max_length=300)
    experience_years: int = Field(default=0, ge=0, le=80)
    license_number: Optional[str] = Field(default=None, max_length=100)
    consultation_fee: Decimal = Field(default=Decimal("0"), ge=0)
    availability_status: AvailabilityStatus = AvailabilityStatus.AVAILABLE
    profile_photo_url: Optional[str] = Field(default=None, max_length=1000)
    bio: Optional[str] = Field(default=None, max_length=5000)

    @field_validator("first_name", "last_name", "phone", "specialization")
    @classmethod
    def strip_required(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field cannot be empty")
        return cleaned

    @field_validator("date_of_birth")
    @classmethod
    def validate_dob(cls, value: Optional[date]) -> Optional[date]:
        if value is None:
            return value
        if value > date.today():
            raise ValueError("Date of birth cannot be in the future")
        if value.year < 1930:
            raise ValueError("Date of birth is unrealistically old")
        return value


class DoctorCreate(DoctorBase):
    pass


class DoctorUpdate(BaseModel):
    first_name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    last_name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    email: Optional[EmailStr] = None
    phone: Optional[str] = Field(default=None, min_length=7, max_length=20)
    gender: Optional[PatientGender] = None
    date_of_birth: Optional[date] = None
    department_id: Optional[UUID] = None
    specialization: Optional[str] = Field(default=None, min_length=1, max_length=150)
    qualification: Optional[str] = Field(default=None, max_length=300)
    experience_years: Optional[int] = Field(default=None, ge=0, le=80)
    license_number: Optional[str] = Field(default=None, max_length=100)
    consultation_fee: Optional[Decimal] = Field(default=None, ge=0)
    availability_status: Optional[AvailabilityStatus] = None
    profile_photo_url: Optional[str] = Field(default=None, max_length=1000)
    bio: Optional[str] = Field(default=None, max_length=5000)


class DepartmentBrief(BaseModel):
    id: UUID
    name: str


class DoctorResponse(DoctorBase, DoctorAIExtensions):
    id: UUID
    doctor_number: str
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    department: Optional[DepartmentBrief] = None

    model_config = {"from_attributes": True}


class DoctorListResponse(BaseModel):
    items: List[DoctorResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class MessageResponse(BaseModel):
    message: str

File: hospital-ai/frontend/src/types/doctor.ts
export type DoctorGender = 'Male' | 'Female' | 'Other';

export type AvailabilityStatus = 'Available' | 'Busy' | 'On Leave';

/** Reserved for future AI agents — currently unused. */
export interface DoctorAIExtensions {
  ai_summary: string | null;
  performance_metrics: unknown | null;
  predicted_workload: unknown | null;
  recommended_schedule: unknown | null;
  /** Reserved for Scheduling Agent — unused. */
  schedule_score?: unknown | null;
}

export interface DepartmentBrief {
  id: string;
  name: string;
}

export interface Doctor extends DoctorAIExtensions {
  id: string;
  doctor_number: string;
  first_name: string;
  last_name: string;
  email: string;
  phone: string;
  gender: DoctorGender;
  date_of_birth: string | null;
  department_id: string | null;
  specialization: string;
  qualification: string | null;
  experience_years: number;
  license_number: string | null;
  consultation_fee: number | string;
  availability_status: AvailabilityStatus;
  profile_photo_url: string | null;
  bio: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  department: DepartmentBrief | null;
}

export interface DoctorFormValues {
  first_name: string;
  last_name: string;
  email: string;
  phone: string;
  gender: DoctorGender | '';
  date_of_birth: string;
  department_id: string;
  specialization: string;
  qualification: string;
  experience_years: string;
  license_number: string;
  consultation_fee: string;
  availability_status: AvailabilityStatus;
  profile_photo_url: string;
  bio: string;
}

export interface DoctorListParams {
  page?: number;
  page_size?: number;
  search?: string;
  department_id?: string;
  availability_status?: AvailabilityStatus | '';
  min_experience?: number | '';
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface DoctorListResponse {
  items: Doctor[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface Department {
  id: string;
  name: string;
  description: string | null;
  floor_number: number | null;
  head_doctor_id: string | null;
  created_at: string;
  updated_at: string;
  doctors_count: number;
  patients_count: number;
  head_doctor_name: string | null;
}

export interface DepartmentFormValues {
  name: string;
  description: string;
  floor_number: string;
  head_doctor_id: string;
}

export interface DepartmentListResponse {
  items: Department[];
  total: number;
}

export interface DoctorAvailability {
  id: string;
  doctor_id: string;
  day_of_week: number;
  start_time: string;
  end_time: string;
  slot_duration: number;
  is_available: boolean;
  created_at: string;
  updated_at: string;
}

export interface AvailabilityFormValues {
  day_of_week: number;
  start_time: string;
  end_time: string;
  slot_duration: string;
  is_available: boolean;
}

export const AVAILABILITY_STATUSES: AvailabilityStatus[] = [
  'Available',
  'Busy',
  'On Leave',
];

export const GENDER_OPTIONS: DoctorGender[] = ['Male', 'Female', 'Other'];

export const DAYS_OF_WEEK = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday',
] as const;

export const EMPTY_DOCTOR_FORM: DoctorFormValues = {
  first_name: '',
  last_name: '',
  email: '',
  phone: '',
  gender: '',
  date_of_birth: '',
  department_id: '',
  specialization: '',
  qualification: '',
  experience_years: '0',
  license_number: '',
  consultation_fee: '0',
  availability_status: 'Available',
  profile_photo_url: '',
  bio: '',
};

export const EMPTY_DEPARTMENT_FORM: DepartmentFormValues = {
  name: '',
  description: '',
  floor_number: '',
  head_doctor_id: '',
};
